FEXCore/Source/Interface/IR/Passes/RegisterAllocationPass.cpp

```diff
@@ -5,1239 +5,676 @@ tags: ir|opts
 $end_info$
 */
 
-#include "Utils/BucketList.h"
-
 #include "Interface/IR/Passes/RegisterAllocationPass.h"
-#include "FEXCore/Core/X86Enums.h"
 #include "Interface/IR/IR.h"
 #include "Interface/IR/IREmitter.h"
 #include "Interface/IR/RegisterAllocationData.h"
 #include "Interface/IR/Passes.h"
-#include <FEXCore/Core/CoreState.h>
 #include <FEXCore/IR/IR.h>
 #include <FEXCore/Utils/LogManager.h>
-#include <FEXCore/Utils/MathUtils.h>
 #include <FEXCore/Utils/Profiler.h>
-#include <FEXCore/Utils/TypeDefines.h>
-#include <FEXCore/fextl/fmt.h>
-#include <FEXCore/fextl/set.h>
-#include <FEXCore/fextl/unordered_map.h>
-#include <FEXCore/fextl/unordered_set.h>
 #include <FEXCore/fextl/vector.h>
+#include <bit>
+#include <cstdint>
 
-#include <FEXHeaderUtils/BitUtils.h>
+using namespace FEXCore;
 
-#include <algorithm>
-#include <cstddef>
-#include <cstdint>
-#include <cstring>
-#include <optional>
-#include <strings.h>
-#include <utility>
+#define foreach_arg(IROp, index, arg) \
```

Comment:
Could we avoid control-flow altering macros like these?

The file-specific idioms they add make reading source code more difficult/surprising when the same thing could be expressed much more clearly with slightly lengthier code. Tooling also tends to handle this poorly (such as debuggers that can't reliably step through code, or IDEs that don't know what line to show when jumping to definitions, etc).

The C++ way to do this equally concisely is to implement a custom iterator (or use C++23 range views), but since both macros don't seem to be used very often we can equally well just use plain index-based loops and initialize Arg in the loop body.

Reply:
Frankly, the C++ ways of doing things have made a mess of FEXCore and have made my work improving FEXCore needlessly challenging. By contrast, the C ways of doing things have served us well in Mesa. I don't think this is a coincidence.

I'm not inclined to make my code worse because it's not C++ enough, I'd rather make the rest of FEXCore more like the competent compilers we have in Mesa. I care more about code being  simple, short, and efficient.. than about it being C++ish.

Reply:
> Frankly, the C++ ways of doing things have made a mess of FEXCore and have made my work improving FEXCore needlessly challenging. By contrast, the C ways of doing things have served us well in Mesa. I don't think this is a coincidence.

It might be worth having a broader discussion about what idioms you've found challenging, but let's focus on this specific instance for the time being.

As mentioned I'm happy with (and actually recommend) a macro-less "C way" too. I wasn't arguing out of pure principle:

> The file-specific idioms they add make reading source code more difficult/surprising when the same thing could be expressed much more clearly with slightly lengthier code.

> Tooling also tends to handle this poorly (such as debuggers that can't reliably step through code, or IDEs that don't know what line to show when jumping to definitions, etc).

To illustrate the first point, it's impossible for me to predict what `foreach_arg(Header, _, Arg) {` will do without jumping between the macro definition and its use several times (even more so since Header is declared without an explicit type). Is this code *simple*? I would consider *local* code that uses standard language syntax simpler, since it clearly and succinctly expresses what variables are being added to the scope.

If the macro were widely used and implemented heavily complicated and repeated logic, I could see some merit in this approach. But when the macros are only used 2/3 times each, what value are they really adding here?

Reply:
I typed a big response to this then Firefox ate it 😭 

Consider some code samples:

* https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/asahi/compiler/agx_opt_empty_else.c
* https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/asahi/compiler/agx_register_allocate.c#L599
* https://github.com/FEX-Emu/FEX/blob/main/FEXCore/Source/Interface/IR/Passes/LongDivideRemovalPass.cpp
* https://github.com/FEX-Emu/FEX/blob/main/FEXCore/Source/Interface/IR/Passes/InlineCallOptimization.cpp

The C code is compact with almost no boilerplate. It expresses complex logic in the language of compiler theory, emphasizing the algorithms and eschewing the specific arrays and lists in the IR. It was easy and fast to write. It is easy and fast to read, facilitated by individually simple helpers that compose to form a rhythm greater than the sum of the parts. It performs well. It will be easy to maintain later. It follows AGX compiler idioms: it is not standard C but it is standard "AGX compiler C".

The C++ code is less compact with piles of pointless class boilerplate. It has fewer expressive helpers, so there is more emphasis on the underlying IR than just the algorithm. Judging by my experiences on this RA and the previous NZCV work, it would have been harder and slower (= more expensive) to write - at least for me. It is - at least for me - harder and slower to read. It is not standard C++ but it is standard "FEXCore C++".

FEXCore has significant performance problems. I want to fix them. Doing so will require rewriting much of FEXCore, redesigning FEXCore, establishing new idioms. Already this RA pass is made far more complex (= slower *and* much more code *and* more expensive to write) because of entrenched FEXCore problems -- Map/Unmap/Remap and Swap2 are just to work around IR deficiencies. The former slows down RA, the latter impedes us from efficiently implementing `pop` instructions. Fixing this is hard ... in part because of the problematic iterators that FEXCore *does* have!

Given the experiences, my inclination is to run as far from C++ as possible. Maybe it's not C++'s fault, but I don't know how to write a good compiler in idiomatic C++ and I do know how to write a good one in non-idiomatic C. I've seen what happens when our team (including me) tries in idiomatic C++, and it isn't working well. Time to try something better. I'm happy to keep `vector` and tuples, but "idiomatic C++" or "idiomatic C" are explicit non-goals for me, in favour of "clean, obvious, fast, compact code" and eventually "idiomatic FEXCore C++".

In my (apparently quixotic) quest to fix FEXCore, I can either attempt a wholesale rewrite or I can improve incrementally. But I can't make incremental progress if the objection is that it doesn't match how we do things today. Well, yeah, that's kind of the point - what we're doing now isn't working but I know something else that does.

Reply:
> I typed a big response to this then Firefox ate it 😭

Having been there and done that, I appreciate you've taken the time to engage in the discussion :)

> Consider some code samples:

Granted, one of these was created by an experienced compiler engineer (you) and uses proven Mesa idioms that passed the test of time, whereas the other was written without this specific expertise and for the longest time focused on getting things running at all more so than on running well.

However, what's clear from your reported experience is that FEX's JIT is not a poster child of good C++. Maybe C++ doesn't even apply all that well to compilers in general. I'm not convinced the latter is the case, but it is what it is regarding the former.

> Given the experiences, my inclination is to run as far from C++ as possible.

If specific patterns fail to apply well to FEX's JIT we should avoid them and stick to simpler approaches. I reject this as an argument for automatically resorting to non-idiomatic code just because the alternative uses C++ however. It helps no one if we fragment FEX into a mix of 20 different coding styles, because I certainly am not convinced that extending Mesa's C-only patterns into FEX's non-JIT modules is a net benefit.

Be that as it may, maybe your approach is a pragmatic solution to dealing with the accumulated debt in the FEX JIT. In this particular instance, this pragmatism leads to a counter-factual decision though:

> Fixing this is hard ... in part because of the problematic iterators that FEXCore does have!

What is the concrete benefit of the `foreach` macros you've added here? The best point I can think of is refactoring, but that's a mood point when - again - we're talking about 5 instances.

I see you've at least dropped the use of `std::tuple` + structured binding from the for-loop to make the macro definitions themselves easier on the eyes. Meanwhile it's still a non-local, file-specific idiom that saves you a minimal number of keystrokes when the alternative of spelling out the for-loops directly code is already simple and and equally performant.

> I can either attempt a wholesale rewrite or I can improve incrementally. But I can't make incremental progress if the objection is that it doesn't match how we do things today

I'm open to incremental improvements, though they should indeed be improvements - otherwise we're just pushing code around for individual convenience. Maybe your ideas can be illustrated more effectively?

Reply:
I'm backing Alyssa's decision here. The paradigms in the code are changing and if C++ is getting in the way then just get rid of the problems for the developer. We are always going to have multiple coding styles in a project, there's no getting around that, we are all squishy humans that write things differently.

We can't stomp on momentum of one person working on new features/implementations because the idioms are closer to C than C++, we'd never get any work done. FEXCore (and more specifically the JIT) is going to be moving more towards C because the people working on the code is more familiar with C, that's completely fine.

Trying to strictly enforce that the codebase must be homogenous in coding style and behaviour is a foolish endeavour and will only lead to burn-out.

Reply:
> I'm backing Alyssa's decision here. The paradigms in the code are changing

As mentioned multiple times, I'm not arguing against change.

> and if C++ is getting in the way then just get rid of the problems for the developer.

This *is* what I argued for. At the same time, C++ is *not* getting in the way of the specific point I commented on.

> Trying to strictly enforce that the codebase must be homogenous in coding style and behaviour is a foolish endeavour and will only lead to burn-out.

No one was trying to enforce a homogenous coding style.

We can agree this general debate isn't useful if it happens on the side of an actual change request. Meanwhile I have raised specific points related to the proposed changes that have not been discussed. Could we stick to the specific change I highlighted about this PR?

Reply:
Dropped the macros, I'm not interested in further litigating this here.